In [66]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
%matplotlib inline

from pandas import DataFrame, Series
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn import metrics

import scikitplot
from scikitplot.metrics import plot_roc_curve as prc
from scikitplot.metrics import plot_lift_curve as plc
from scikitplot.metrics import plot_cumulative_gain as pcg
from sklearn.metrics import classification_report

# Ignore useless warnings
import warnings
warnings.filterwarnings("ignore")

In [67]:
test = pd.read_csv("general_data.csv")

In [68]:
test.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [69]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [70]:
test.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [71]:
test.astype(bool).sum(axis=0)

Age                        4410
Attrition                  4410
BusinessTravel             4410
Department                 4410
DistanceFromHome           4410
Education                  4410
EducationField             4410
EmployeeCount              4410
EmployeeID                 4410
Gender                     4410
JobLevel                   4410
JobRole                    4410
MaritalStatus              4410
MonthlyIncome              4410
NumCompaniesWorked         3824
Over18                     4410
PercentSalaryHike          4410
StandardHours              4410
StockOptionLevel           2517
TotalWorkingYears          4377
TrainingTimesLastYear      4248
YearsAtCompany             4278
YearsSinceLastPromotion    2667
YearsWithCurrManager       3621
dtype: int64

In [72]:
test['EmployeeCount'].value_counts()

1    4410
Name: EmployeeCount, dtype: int64

Ils sont tous à 1 donc osef

Après concertation, nous avons décidé de supprimer les données Over18, EmployeeCount, MaritalStatus et Gender. 

In [73]:
test = test.drop(columns=['EmployeeCount','Over18','Gender','MaritalStatus'])

Attrition : yes -> l'employé est parti en 2016 / no -> il est toujours dans l'entreprise

In [74]:
test['Attrition'].value_counts()

No     3699
Yes     711
Name: Attrition, dtype: int64

On cherche à savoir pourquoi ces 711 employés ont quitté l'entreprise, nous allons donc uniquement nous concentrer sur les données où l'Attrition est 'yes'.

In [75]:
test.set_index('Attrition')

Age     BusinessTravel              Department  DistanceFromHome  \
Attrition                                                                     
No          51      Travel_Rarely                   Sales                 6   
Yes         31  Travel_Frequently  Research & Development                10   
No          32  Travel_Frequently  Research & Development                17   
No          38         Non-Travel  Research & Development                 2   
No          32      Travel_Rarely  Research & Development                10   
...        ...                ...                     ...               ...   
No          42      Travel_Rarely  Research & Development                 5   
No          29      Travel_Rarely  Research & Development                 2   
No          25      Travel_Rarely  Research & Development                25   
No          42      Travel_Rarely                   Sales                18   
No          40      Travel_Rarely  Research & Development                28   

           Education EducationField  EmployeeID  JobLevel  \
Attrition                                                   
No                 2  Life Sciences           1         1   
Yes                1  Life Sciences           2         1   
No                 4          Other           3         4   
No                 5  Life Sciences           4         3   
No                 1        Medical           5         1   
...              ...            ...         ...       ...   
No                 4        Medical        4406         1   
No                 4        Medical        4407         1   
No                 2  Life Sciences        4408         2   
No                 2        Medical        4409         1   
No                 3        Medical        4410         2   

                             JobRole  MonthlyIncome  NumCompaniesWorked  \
Attrition                                                                 
No         Healthcare Representative         131160                 1.0   
Yes               Research Scientist          41890                 0.0   
No                   Sales Executive         193280                 1.0   
No                   Human Resources          83210                 3.0   
No                   Sales Executive          23420                 4.0   
...                              ...            ...                 ...   
No                Research Scientist          60290                 3.0   
No             Laboratory Technician          26790                 2.0   
No                   Sales Executive          37020                 0.0   
No             Laboratory Technician          23980                 0.0   
No             Laboratory Technician          54680                 0.0   

           PercentSalaryHike  StandardHours  StockOptionLevel  \
Attrition                                                       
No                        11              8                 0   
Yes                       23              8                 1   
No                        15              8                 3   
No                        11              8                 3   
No                        12              8                 2   
...                      ...            ...               ...   
No                        17              8                 1   
No                        15              8                 0   
No                        20              8                 0   
No                        14              8                 1   
No                        12              8                 0   

           TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
Attrition                                                             
No                       1.0                      6               1   
Yes                      6.0                      3               5   
No                       5.0                      2               5   
No          

In [77]:
test_without_no = test.drop(test[test.Attrition == 'No'].index)
test_without_no

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4381   29       Yes      Travel_Rarely  Research & Development   
4386   33       Yes      Travel_Rarely                   Sales   
4388   33       Yes      Travel_Rarely                   Sales   
4391   32       Yes      Travel_Rarely                   Sales   
4402   37       Yes  Travel_Frequently                   Sales   

      DistanceFromHome  Education EducationField  EmployeeID  JobLevel  \
1                   10          1  Life Sciences           2         1   
6                   11          2        Medical           7         2   
13                   1          1        Medical          14         1   
28                   1          2        Medical          29         2   
30                   4          3        Medical          31         3   
...                ...        ...            ...         ...       ...   
4381                 7          1  Life Sciences        4382         1   
4386                11          4      Marketing        4387         1   
4388                 1          3  Life Sciences        4389         2   
4391                23          1  Life Sciences        4392         3   
4402                 2          3      Marketing        4403         1   

                        JobRole  MonthlyIncome  NumCompaniesWorked  \
1            Research Scientist          41890                 0.0   
6               Sales Executive          58130                 2.0   
13           Research Scientist          57620                 1.0   
28           Research Scientist         103330                 3.0   
30           Research Scientist          68540                 2.0   
...                         ...            ...                 ...   
4381         Research Scientist          21800                 1.0   
4386         Research Scientist          71400                 5.0   
4388                    Manager          51470                 7.0   
4391  Healthcare Representative          24680                 0.0   
4402      Laboratory Technician          40010                 6.0   

      PercentSalaryHike  StandardHours  StockOptionLevel  TotalWorkingYears  \
1                    23              8                 1                6.0   
6                    20              8                 1                5.0   
13                   11              8                 2               10.0   
28                   14              8                 1               19.0   
30                   11              8                 0                5.0   
...                 ...            ...               ...                ...   
4381                 21              8                 0                4.0   
4386                 21              8                 0                8.0   
4388                 11              8                 0               13.0   
4391                 11              8                 0                4.0   
4402                 11              8                 1               17.0   

      TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
1                         3               5                        1   
6                         2               0                        0   
13                        4              10                        9   
28                        2               1                        0   
30                        5               3                        0   
...                     ...             ...                      ...   
4381                      2   

In [82]:
test_without_no['Department'].value_counts()

Research & Development    453
Sales                     201
Human Resources            57
Name: Department, dtype: int64

In [83]:
test_RandD = test_without_no[test_without_no.Department == 'Research & Development']
test_RandD

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4344   35       Yes      Travel_Rarely  Research & Development   
4362   32       Yes      Travel_Rarely  Research & Development   
4366   30       Yes      Travel_Rarely  Research & Development   
4369   29       Yes      Travel_Rarely  Research & Development   
4381   29       Yes      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeID  JobLevel  \
1                   10          1  Life Sciences           2         1   
6                   11          2        Medical           7         2   
13                   1          1        Medical          14         1   
28                   1          2        Medical          29         2   
30                   4          3        Medical          31         3   
...                ...        ...            ...         ...       ...   
4344                 2          4  Life Sciences        4345         2   
4362                 5          3        Medical        4363         1   
4366                10          4  Life Sciences        4367         2   
4369                 2          3  Life Sciences        4370         1   
4381                 7          1  Life Sciences        4382         1   

                        JobRole  MonthlyIncome  NumCompaniesWorked  \
1            Research Scientist          41890                 0.0   
6               Sales Executive          58130                 2.0   
13           Research Scientist          57620                 1.0   
28           Research Scientist         103330                 3.0   
30           Research Scientist          68540                 2.0   
...                         ...            ...                 ...   
4344          Research Director         108830                 3.0   
4362  Healthcare Representative          34770                 1.0   
4366         Research Scientist          55620                 5.0   
4369            Sales Executive          49780                 0.0   
4381         Research Scientist          21800                 1.0   

      PercentSalaryHike  StandardHours  StockOptionLevel  TotalWorkingYears  \
1                    23              8                 1                6.0   
6                    20              8                 1                5.0   
13                   11              8                 2               10.0   
28                   14              8                 1               19.0   
30                   11              8                 0                5.0   
...                 ...            ...               ...                ...   
4344                 12              8                 0                8.0   
4362                 11              8                 0               14.0   
4366                 13              8                 0                8.0   
4369                 13              8                 0                7.0   
4381                 21              8                 0                4.0   

      TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
1                         3               5                        1   
6                         2               0                        0   
13                        4              10                        9   
28                        2               1                        0   
30                        5               3                        0   
...                     ...             ...                      ...   
4344                      3   

In [84]:
test_Sales = test_without_no[test_without_no.Department == 'Sales']
test_Sales

Age Attrition     BusinessTravel Department  DistanceFromHome  \
124    26       Yes  Travel_Frequently      Sales                14   
127    30       Yes  Travel_Frequently      Sales                 3   
178    33       Yes      Travel_Rarely      Sales                 5   
182    34       Yes  Travel_Frequently      Sales                17   
205    26       Yes      Travel_Rarely      Sales                 1   
...   ...       ...                ...        ...               ...   
4368   30       Yes  Travel_Frequently      Sales                 3   
4386   33       Yes      Travel_Rarely      Sales                11   
4388   33       Yes      Travel_Rarely      Sales                 1   
4391   32       Yes      Travel_Rarely      Sales                23   
4402   37       Yes  Travel_Frequently      Sales                 2   

      Education EducationField  EmployeeID  JobLevel  \
124           3  Life Sciences         125         2   
127           2      Marketing         128         3   
178           4      Marketing         179         2   
182           1      Marketing         183         1   
205           3      Marketing         206         2   
...         ...            ...         ...       ...   
4368          2        Medical        4369         2   
4386          4      Marketing        4387         1   
4388          3  Life Sciences        4389         2   
4391          1  Life Sciences        4392         3   
4402          3      Marketing        4403         1   

                        JobRole  MonthlyIncome  NumCompaniesWorked  \
124                     Manager          75100                 0.0   
127          Research Scientist          11290                 9.0   
178           Research Director          66520                 7.0   
182           Research Director          20010                 9.0   
205      Manufacturing Director          24260                 1.0   
...                         ...            ...                 ...   
4368  Healthcare Representative          28620                 5.0   
4386         Research Scientist          71400                 5.0   
4388                    Manager          51470                 7.0   
4391  Healthcare Representative          24680                 0.0   
4402      Laboratory Technician          40010                 6.0   

      PercentSalaryHike  StandardHours  StockOptionLevel  TotalWorkingYears  \
124                  18              8                 3                7.0   
127                  19              8                 0                3.0   
178                  25              8                 0                8.0   
182                  22              8                 1               11.0   
205                  22              8                 0                1.0   
...                 ...            ...               ...                ...   
4368                 14              8                 0                9.0   
4386                 21              8                 0                8.0   
4388                 11              8                 0               13.0   
4391                 11              8                 0                4.0   
4402                 11              8                 1               17.0   

      TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
124                       3               6                        0   
127                       2               1                        0   
178                       5               4                        1   
182                       5               3                        0   
205                       3               1                        0   
...                     ...             ...                      ...   
4368                      2               6                        0   
4386                      2               5                        0   
4388                      2               

In [85]:
test_HR = test_without_no[test_without_no.Department == 'Human Resources']
test_HR

Age Attrition     BusinessTravel       Department  DistanceFromHome  \
105    33       Yes      Travel_Rarely  Human Resources                28   
112    21       Yes      Travel_Rarely  Human Resources                10   
310    55       Yes      Travel_Rarely  Human Resources                 7   
350    55       Yes      Travel_Rarely  Human Resources                16   
453    29       Yes      Travel_Rarely  Human Resources                 3   
538    25       Yes      Travel_Rarely  Human Resources                25   
599    44       Yes      Travel_Rarely  Human Resources                 6   
613    24       Yes      Travel_Rarely  Human Resources                28   
655    45       Yes      Travel_Rarely  Human Resources                12   
878    28       Yes      Travel_Rarely  Human Resources                 1   
943    23       Yes      Travel_Rarely  Human Resources                 6   
999    31       Yes      Travel_Rarely  Human Resources                 9   
1039   26       Yes      Travel_Rarely  Human Resources                 2   
1096   56       Yes      Travel_Rarely  Human Resources                 1   
1107   29       Yes  Travel_Frequently  Human Resources                 2   
1245   39       Yes      Travel_Rarely  Human Resources                 1   
1246   24       Yes      Travel_Rarely  Human Resources                 2   
1312   37       Yes      Travel_Rarely  Human Resources                 9   
1323   34       Yes  Travel_Frequently  Human Resources                 4   
1575   33       Yes      Travel_Rarely  Human Resources                28   
1582   21       Yes      Travel_Rarely  Human Resources                10   
1780   55       Yes      Travel_Rarely  Human Resources                 7   
1820   55       Yes      Travel_Rarely  Human Resources                16   
1923   29       Yes      Travel_Rarely  Human Resources                 3   
2008   25       Yes      Travel_Rarely  Human Resources                25   
2069   44       Yes      Travel_Rarely  Human Resources                 6   
2083   24       Yes      Travel_Rarely  Human Resources                28   
2125   45       Yes      Travel_Rarely  Human Resources                12   
2348   28       Yes      Travel_Rarely  Human Resources                 1   
2413   23       Yes      Travel_Rarely  Human Resources                 6   
2469   31       Yes      Travel_Rarely  Human Resources                 9   
2509   26       Yes      Travel_Rarely  Human Resources                 2   
2566   56       Yes      Travel_Rarely  Human Resources                 1   
2577   29       Yes  Travel_Frequently  Human Resources                 2   
2715   39       Yes      Travel_Rarely  Human Resources                 1   
2716   24       Yes      Travel_Rarely  Human Resources                 2   
2782   37       Yes      Travel_Rarely  Human Resources                 9   
2793   34       Yes  Travel_Frequently  Human Resources                 4   
3045   33       Yes      Travel_Rarely  Human Resources                28   
3052   21       Yes      Travel_Rarely  Human Resources                10   
3250   55       Yes      Travel_Rarely  Human Resources                 7   
3290   55       Yes      Travel_Rarely  Human Resources                16   
3393   29       Yes      Travel_Rarely  Human Resources                 3   
3478   25       Yes      Travel_Rarely  Human Resources                25   
3539   44       Yes      Travel_Rarely  Human Resources                 6   
3553   24       Yes      Travel_Rarely  Human Resources                28   
3595   45       Yes      Travel_Rarely  Human Resources                12   
3818   28       Yes      Travel_Rarely  Human Resources                 1   
3883   23       Yes      Travel_Rarely  Human Resources                 6   
3939   31       Yes      Travel_Rarely  Human Resources                 9   
3979   26       Yes      Travel_Rarely  Human Resources                 2   
40

Le nombre de départs d'employés travaillant aux Ressources Humaines ne représentant que 8.02% des départs, nous faisons le choix de les ignorer pour le moment.